src/wp-includes/canonical.php

```diff
@@ -229,49 +276,69 @@ function redirect_canonical( $requested_url = null, $do_redirect = true ) {
+							$tax_obj = get_taxonomy( $obj->taxonomy );
+							if ( false !== $tax_obj->query_var ) {
+								$qv_remove[] = $tax_obj->query_var;
+							}
 						}
-					}
 
-					$rewrite_vars = array_diff( array_keys( $wp_query->query ), array_keys( $_GET ) );
+						$rewrite_vars = array_diff( array_keys( $wp_query->query ), array_keys( $_GET ) );
 
-					if ( ! array_diff( $rewrite_vars, array_keys( $_GET ) ) ) { // Check to see if all the Query vars are coming from the rewrite, none are set via $_GET
-						$redirect['query'] = remove_query_arg( $qv_remove, $redirect['query'] ); //Remove all of the per-tax qv's
+						// Check to see if all the query vars are coming from the rewrite, none are set via $_GET.
+						if ( ! array_diff( $rewrite_vars, array_keys( $_GET ) ) ) {
+							// Remove all of the per-tax query vars.
+							$redirect['query'] = remove_query_arg( $qv_remove, $redirect['query'] );
 
-						// Create the destination url for this taxonomy
-						$tax_url = parse_url( $tax_url );
-						if ( ! empty( $tax_url['query'] ) ) { // Taxonomy accessible via ?taxonomy=..&term=.. or any custom qv..
-							parse_str( $tax_url['query'], $query_vars );
-							$redirect['query'] = add_query_arg( $query_vars, $redirect['query'] );
-						} else { // Taxonomy is accessible via a "pretty-URL"
-							$redirect['path'] = $tax_url['path'];
-						}
-					} else { // Some query vars are set via $_GET. Unset those from $_GET that exist via the rewrite
-						foreach ( $qv_remove as $_qv ) {
-							if ( isset( $rewrite_vars[ $_qv ] ) ) {
-								$redirect['query'] = remove_query_arg( $_qv, $redirect['query'] );
+							// Create the destination URL for this taxonomy.
+							$tax_url = parse_url( $tax_url );
+
+							if ( ! empty( $tax_url['query'] ) ) {
+								// Taxonomy accessible via ?taxonomy=...&term=... or any custom query var.
+								parse_str( $tax_url['query'], $query_vars );
+								$redirect['query'] = add_query_arg( $query_vars, $redirect['query'] );
+							} else {
+								// Taxonomy is accessible via a "pretty URL".
+								$redirect['path'] = $tax_url['path'];
+							}
+						} else {
+							// Some query vars are set via $_GET. Unset those from $_GET that exist via the rewrite.
+							foreach ( $qv_remove as $_qv ) {
+								if ( isset( $rewrite_vars[ $_qv ] ) ) {
+									$redirect['query'] = remove_query_arg( $_qv, $redirect['query'] );
+								}
 							}
 						}
 					}
 				}
 			}
-		} elseif ( is_single() && strpos( $wp_rewrite->permalink_structure, '%category%' ) !== false && $cat = get_query_var( 'category_name' ) ) {
-			$category = get_category_by_path( $cat );
-			if ( ( ! $category || is_wp_error( $category ) ) || ! has_term( $category->term_id, 'category', $wp_query->get_queried_object_id() ) ) {
-				$redirect_url = get_permalink( $wp_query->get_queried_object_id() );
+		} elseif ( is_single() && strpos( $wp_rewrite->permalink_structure, '%category%' ) !== false ) {
```

Comment:
Yoda needed `false  !== strpos( $wp_rewrite->permalink_structure, '%category%' )`

Reply:
Nice - that statement isn't picked up by the Yoda sniff - issue reported upstream.